### ISSUE TITLE
Atoum don't keep mock defined in beforeTestMethod

### ISSUE THREAD

**Issue by Guillaume-Rossignol** [atoum/atoum#426]
Since commit b0fb4c50 this test fail : 

```
<?php
namespace Cheznous\tests\units;
use \atoum;
class Test
{
    public function put()
    {
        return true;
    }
}

class Image extends atoum
{
    protected $mocked_storage = null;

    public function beforeTestMethod($method)
    {
        $this->mocked_storage = new \mock\Cheznous\tests\units\Test;
        $this->mocked_storage->getMockController()->put = function() {return 10;};
    }

    public function testMyTest()
    {
        $this->variable($this->mocked_storage->put('arg1','arg2'))->isIdenticalTo(10);//fail with b0fb, return true
    }

}
```
If we call manually beforeTestMethod the test success

```
class Image extends atoum
{
    protected $mocked_storage = null;

    public function beforeTestMethod($method)
    {
        $this->mocked_storage = new \mock\Cheznous\tests\units\Test;
        $this->mocked_storage->getMockController()->put = function() {return 10;};
    }

    public function testMyTest()
    {
        $this->beforeTestMethod("testMyTest");
        $this->variable($this->mocked_storage->put('arg1','arg2'))->isIdenticalTo(10);//success
    }

}
```

**Comment by mikaelrandy** [comment]
In my mind, it's not a good practice to always defined mocks, especially if all of your test methods not use it.

Why do not define a protected method to define your mock ?

```php
class Image extends atoum
{
    protected function getStorageMock()
    {
        $storage = new \mock\Cheznous\tests\units\Test;
        $storage->getMockController()->put = function() {return 10;};

        return $storage;
    }

    public function testMyTest()
    {
        $storage = $this->getMockedStorage();
        $this->variable($storage->put('arg1','arg2'))->isIdenticalTo(10);//success
    }

}
```

**Comment by b-viguier** [comment]
@mikaelrandy I agree with the workaround and the guideline. Nevertheless, IMHO, it's a breaking change...

**Comment by mikaelrandy** [comment]
It's linked to #420

[Now atoum is versionned](http://jubianchi.fr/atoum-release.htm), you can use the "~1.2" version

**Comment by mageekguy** [comment]
Think that it's possible to remove this BC break.
@b-viguier can you edit file `classes/test.php` and moving line 1134~1137 to line 1094, and running your test suite?

**Comment by Guillaume-Rossignol** [comment]
@mageekguy Our tests suite works with this modification.

And we have updated composer.json to use stable version \o/

**Comment by Hywan** [comment]
So we have a regression, we have to create a new release. @mikaelrandy: Can you do this please, I don't have time right now?

**Comment by mageekguy** [comment]
@Hywan read carefully last comment of @b-viguier...

**Comment by b-viguier** [comment]
@mageekguy You certainly mean *last comment of @Guillaume-Rossignol * :)

Your patch seems to solve the regression, then up to atoum's team to decide what to do :)

**Comment by jubianchi** [comment]
@mageekguy @Guillaume-Rossignol @b-viguier If applying this patch works well with both test suites I can open a PR and eventually ship a new release ASAP.

Just let me know ;)

/cc @Hywan

**Comment by mikaelrandy** [comment]
@jubianchi as #420 is not yet released, if an other PR correct the regression, i think there is no emergency to ship a new release.

**Comment by jubianchi** [comment]
@mikaelrandy you are right and this is perfect ;) We'll be able to ship a new release with a new "feature" and no regression.

:+1:

**Comment by b-viguier** [comment]
:+1:

@jubianchi I think the PR could/should contain a unit test to prevent this regression

**Comment by jubianchi** [comment]
@b-viguier of course ;)

**Comment by mageekguy** [comment]
Good luck...

**Comment by jubianchi** [comment]
Should be fixed with 7d1b885.
